ng-dev/misc/new-main-branch/cli.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {assertValidGithubConfig, getConfig} from '../../utils/config';
+import {error, green, info, promptConfirm, red, warn, yellow} from '../../utils/console';
+import {getRemotesForRepo, isAngularOwnedRemote} from './remotes';
+
+import {CommandModule} from 'yargs';
+import {GitClient} from '../../utils/git/git-client';
+import {promptForRemoteForkUpdate} from './remote-fork-update';
+
+/**
+ * Migration command that performs local changes to account for an upstream
+ * branch rename from `master` to `main`. More details can be found here:
+ *
+ * https://docs.google.com/document/d/1nqb94eSIcGuPC0M9Rv7-IeqOiJKzsMnuistHZNeUmAg.
+ */
+async function handler() {
+  const git = GitClient.get();
+  const config = getConfig([assertValidGithubConfig]);
+  const repoSlug = `${config.github.owner}/${config.github.name}`;
+
+  if (config.github.mainBranchName !== 'main') {
+    error(red('Current project is not part of the default branch renaming.'));
+    return;
+  }
+
+  if (git.hasUncommittedChanges()) {
+    error(red('There are uncommitted changes. Unable to switch to new main branch. Aborting..'));
+    return;
+  }
+
+  if (!hasLocalBranch(git, 'master')) {
+    error(red('Local repository does not have a local branch named `master`. Aborting..'));
+    return;
+  }
+
+  if (hasLocalBranch(git, 'main')) {
+    error(red('Local repository already has a branch named `main`. Aborting..'));
+    return;
+  }
+
+  const remotes = getRemotesForRepo(git);
+  const angularRemotes = Array.from(remotes.entries()).filter((r) => isAngularOwnedRemote(r[1]));
+  const angularRemoteNames = angularRemotes.map(([name]) => name);
+
+  // Usually we expect only a single remote pointing to the Angular-owned Github
+  // repository. If there are more, we will just assume the first one to be the primary.
+  const primaryRemoteName = angularRemoteNames[0];
+
+  if (angularRemoteNames.length === 0) {
+    warn(yellow(`Found no remote in repository that points to the \`${repoSlug}\` repository.`));
+  }
+
+  info('The following steps will be performed:');
+  if (angularRemoteNames.length) {
+    info(`  → Remotes (${angularRemoteNames.join(`, `)}) are refreshed.`);
+  }
+  info(`  → The \`main\` branch is fetched from \`${repoSlug}\`.`);
+  info('  → The new `main` branch is checked out.');
+  if (primaryRemoteName) {
+    info(`  → The new \`main\` branch is linked to the \`${primaryRemoteName}\` remote.`);
+  }
+  info('  → The old `master` branch is deleted or renamed (you will be prompted).');
+  info('  → Remote references to `master` branches in the index are removed.');
+  info('');
+
+  if (!(await promptConfirm('Do you want to continue?'))) {
+    return;
+  }
+
+  // Refreshing Angular-owned remotes.
+  git.run(['fetch', ...angularRemoteNames]);
```

Comment:
We should make sure that any git outputs from these commands go to `debug` so that users don't see unexpected/unnecessary output in the terminal.

Additionally, should we put a spinner to express what we are doing while it happens?

Reply:
Updated to fix this, and pushed an additional `GitClient` fix/change